### ISSUE TITLE
Mapping Tool Not Scaling Correctly

### ISSUE THREAD

**Issue by nickkeane** [dosbox-staging/dosbox-staging#2237]
### Are you using the latest Dosbox-Staging Version?

- [X] I have checked releases and am using the latest release.

### Different version than latest?

0.80.1

### What Operating System are you using?

Windows 10

### Have you checked that no other similar issue already exists?

- [X] I have searched and not found similar issues.

### A clear and concise description of what the bug is.

The Mapper Tool (Ctrl+F1) is not scaling to the correct dimensions. When running DOSBox Staging in fullscreen or windowed mode, the Mapper Tool pops up in a very tiny window, and the mouse is not working correctly (the x,y pos seems to be scaled improperly on the mouse position so that it isn't clicking the buttons correctly). It may have something to do with my laptop being 2560x1440, but I'm not sure.

### Steps to reproduce the behaviour.

Explain how to reproduce
1. Run DOSBox Staging
2. Press Ctrl+F1 to open the mapping tool
3. See the window and try to click some buttons directly

### Your configuration

```ini
https://gist.github.com/nickkeane/eed5e0d23e633fd65b3fbb5515d733b1
```

### Code of Conduct & Contributing Guidelines

- [X] Yes, I agree.

**Comment by kcgen** [comment]
Thanks @NicknineTheEagle. 

Is your screen's magnification set to 100%?

The mapper's elements are placed using unscaled coordinates, so I guess that should be add.

Or maybe when the mapper is open, the mouse needs to switch to DPI-non-aware mode.

Ideally on every SDL window mode change, the mouse's coordinate system and DPI would be updated.

**Comment by nickkeane** [comment]
I don't use a magnification tool, so I would assume so.

**Comment by GranMinigun** [comment]
Mapper window uses outdated surface output and isn't DPI aware. I think moving to texture output resolved that problem in my branch, but it'll take some time to go into release.

**Comment by finalpatch** [comment]
I want to add that the debugger window is affected by the same problem.  If I set scaling to 200% then only the top left part of the window is rendered, and the other 3/4 shows rubbish.

**Comment by NicknineTheEagle** [comment]
> Thanks @NicknineTheEagle. 
> 
> Is your screen's magnification set to 100%?
> 
> The mapper's elements are placed using unscaled coordinates, so I guess that should be updated.
> 
> Or maybe when the mapper is open, the mouse needs to switch to DPI-non-aware mode.
> 
> Ideally on every SDL window mode change, the mouse's coordinate system and DPI/scaling would be updated.

Wrong Nick. :P

**Comment by finalpatch** [comment]
I believe both the mapper and the debugger will be fixed once this SDL change reaches upstream:
https://github.com/libsdl-org/SDL/pull/7069

I was also able to fix the mapper by performing the final blit to screen with a sdl renderer instead of surface.

**Comment by finalpatch** [comment]
My renderer diff that fixes the mapper for me on windows 11 with 200% scaling
```
modified   src/gui/sdl_mapper.cpp
@@ -1386,6 +1386,8 @@ static struct CMapper {
 	SDL_Surface *draw_surface_nonpaletted = nullptr; // Needed for SDL_BlitScaled
 	SDL_Surface *surface = nullptr;
 	SDL_Surface *draw_surface = nullptr;
+    SDL_Renderer *renderer = nullptr;
+    SDL_Texture *texture = nullptr;
 	bool exit = false;
 	CEvent *aevent = nullptr;  // Active Event
 	CBind *abind = nullptr;    // Active Bind
@@ -1988,9 +1990,9 @@ static void DrawButtons() {
 	}
 	// We can't just use SDL_BlitScaled (say for Android) in one step
 	SDL_BlitSurface(mapper.draw_surface, NULL, mapper.draw_surface_nonpaletted, NULL);
-	SDL_BlitScaled(mapper.draw_surface_nonpaletted, NULL, mapper.surface, &mapper.draw_rect);
-	//SDL_BlitSurface(mapper.draw_surface, NULL, mapper.surface, NULL);
-	SDL_UpdateWindowSurface(mapper.window);
+    SDL_UpdateTexture(mapper.texture, NULL, mapper.draw_surface_nonpaletted->pixels, mapper.draw_surface_nonpaletted->pitch);
+    SDL_RenderCopy(mapper.renderer, mapper.texture, NULL, NULL);
+    SDL_RenderPresent(mapper.renderer);
 }
 
 static CKeyEvent * AddKeyButtonEvent(Bitu x,Bitu y,Bitu dx,Bitu dy,char const * const title,char const * const entry,KBD_KEYS key) {
@@ -2992,7 +2994,9 @@ void MAPPER_DisplayUI() {
 	/* Set some palette entries */
 	mapper.draw_surface=SDL_CreateRGBSurface(0,640,480,8,0,0,0,0);
 	// Needed for SDL_BlitScaled
-	mapper.draw_surface_nonpaletted=SDL_CreateRGBSurface(0,640,480,32,0x0000ff00,0x00ff0000,0xff000000,0);
+	mapper.draw_surface_nonpaletted=SDL_CreateRGBSurface(0,640,480,32,0x00ff0000,0x00ff00,0x0000ff,0);
+    mapper.renderer = SDL_CreateRenderer(mapper.window, -1, SDL_RENDERER_ACCELERATED);
+    mapper.texture = SDL_CreateTextureFromSurface(mapper.renderer, mapper.draw_surface_nonpaletted);
 	mapper.draw_rect=GFX_GetSDLSurfaceSubwindowDims(640,480);
 	// Sorry, but SDL_SetSurfacePalette requires a full palette.
 	SDL_Palette *sdl2_map_pal_ptr = SDL_AllocPalette(256);
@@ -3014,13 +3018,15 @@ void MAPPER_DisplayUI() {
 			mapper.redraw = false;
 			DrawButtons();
 		} else {
-			SDL_UpdateWindowSurface(mapper.window);
+            SDL_RenderPresent(mapper.renderer);
 		}
 		BIND_MappingEvents();
 		Delay(1);
 	}
 	/* ONE SHOULD NOT FORGET TO DO THIS!
 	Unless a memory leak is desired... */
+    SDL_DestroyTexture(mapper.texture);
+    SDL_DestroyRenderer(mapper.renderer);
 	SDL_FreeSurface(mapper.draw_surface);
 	SDL_FreeSurface(mapper.draw_surface_nonpaletted);
 	SDL_FreePalette(sdl2_map_pal_ptr);
```

**Comment by GranMinigun** [comment]
You want to check `gm/sdlpainkiller-1` branch. It already ports mapper to SDL renderer. Debugger will likely require some modifications to pdcurses.

**Comment by GranMinigun** [comment]
Should be fixed by #2763.
